source/extensions/filters/http/ext_proc/processor_state.cc

```diff
@@ -142,146 +142,183 @@ ProcessorState::getCallbackStateAfterHeaderResp(const CommonResponse& common_res
+absl::Status ProcessorState::handleHeaderContinue(const HeadersResponse& response) {
+  if (no_body_) {
+    // Fall through if there was never a body in the first place.
+    ENVOY_STREAM_LOG(debug, "The message had no body", *filter_callbacks_);
+  } else if (complete_body_available_ && body_mode_ != ProcessingMode::NONE) {
+    return handleCompleteBodyAvailable(response);
+  } else if (body_mode_ == ProcessingMode::BUFFERED) {
+    // Here, we're not ready to continue processing because then
+    // we won't be able to modify the headers any more, so do nothing and
+    // let the doData callback handle body chunks until the end is reached.
+    clearWatermark();
+    filter_.onProcessHeadersResponse(response, absl::OkStatus(), trafficDirection());
+    return absl::OkStatus();
+  } else if (body_mode_ == ProcessingMode::STREAMED) {
+    sendBufferedDataInStreamedMode(false);
+    filter_.onProcessHeadersResponse(response, absl::OkStatus(), trafficDirection());
+    continueIfNecessary();
+    return absl::OkStatus();
+  } else if (body_mode_ == ProcessingMode::FULL_DUPLEX_STREAMED) {
+    // There is no buffered data in this mode.
     filter_.onProcessHeadersResponse(response, absl::OkStatus(), trafficDirection());
     continueIfNecessary();
+    return absl::OkStatus();
+  } else if (body_mode_ == ProcessingMode::BUFFERED_PARTIAL) {
+    return handleBufferedPartialMode(response);
+  }
+
+  return handleTrailersAndCleanup(response);
+}
+
+absl::Status ProcessorState::handleCompleteBodyAvailable(const HeadersResponse& response) {
+  if (callback_state_ == CallbackState::Idle) {
+    // If we get here, then all the body data came in before the header message
+    // was complete, and the server wants the body. It doesn't matter whether the
+    // processing mode is buffered, streamed, or partially buffered.
+    if (bufferedData()) {
+      // Get here, no_body_ = false, and complete_body_available_ = true, the end_stream
+      // flag of decodeData() can be determined by whether the trailers are received.
+      // Also, bufferedData() is not nullptr means decodeData() is called, even though
+      // the data can be an empty chunk.
+      auto req = filter_.setupBodyChunk(*this, *bufferedData(), trailers_ == nullptr);
+      filter_.sendBodyChunk(*this, ProcessorState::CallbackState::BufferedBodyCallback, req);
+      clearWatermark();
+      filter_.onProcessHeadersResponse(response, absl::OkStatus(), trafficDirection());
+      return absl::OkStatus();
+    }
+  } else {
+    // StreamedBodyCallback state. There is pending body response.
+    // Check whether there is buffered data. If there is, send them.
+    // Do not continue filter chain here so the pending body response have chance to be
+    // served.
+    sendBufferedDataInStreamedMode(trailers_ == nullptr);
+    filter_.onProcessHeadersResponse(response, absl::OkStatus(), trafficDirection());
+    return absl::OkStatus();
+  }
+
+  return handleTrailersAndCleanup(response);
```

Comment:
Moving this return to after line 260 is cleaner as it will only be called if (bufferedData(), i.e. line 250  returns false.

Reply:
Done.